File: 442-Find-All-Duplicates-in-an-Array/py/main.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        duplicates = []
    
        for num in nums:
            # Get the absolute value of the number since we might have made it negative previously
            index = abs(num) - 1
            
            # If the value at index is negative, it means we've already seen this number before
            if nums[index] < 0:
                duplicates.append(index + 1)
            else:
                # Mark the number as visited by making its value negative
                nums[index] = -nums[index]
        
        return duplicates

File: 442-Find-All-Duplicates-in-an-Array/rs/main.rs
pub fn find_duplicates(mut nums: Vec<i32>) -> Vec<i32> {
    let mut dup_arr: Vec<i32> = vec![];
    let mut num_arr = nums.clone();

    for num in 0..nums.len() {
        let index = num_arr[num].abs() as usize - 1;

        if num_arr[index] < 0 {
            dup_arr.push(index as i32 + 1);
        }
        num_arr[index] *= -1;
    }

    dup_arr
}


fn main() {
        let nums: Vec<i32> = vec![4,3,2,7,8,2,3,1];
        find_duplicates(nums);
}
